core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java

```diff
@@ -0,0 +1,190 @@
+  private void startMonitorExecutorService() {
+    Runnable print =
+        () -> logger.info("[MONITOR] Timestamp={}, Metrics={}", Instant.now(), getMetrics());
+
+    monitorExecutorService.execute(
+        () -> {
+          while (!monitorExecutorService.isShutdown()) {
+            print.run();
+            Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
+          }
+          print.run();
+        });
+  }
+
+  /**
+   * Set an emitter which contains implementation to emit values. Ideally, this should be passed to
+   * the constructor, but the instantiation timings of {@link GroupCommitter} and {@link Emittable}
+   * can be different. That's why this API exists.
+   *
+   * @param emitter An emitter.
+   */
+  public void setEmitter(Emittable<EMIT_KEY, V> emitter) {
+    groupManager.setEmitter(emitter);
+  }
+
+  /**
+   * Reserves a new slot in the current {@link NormalGroup}. The slot may be moved to a {@link
+   * DelayedGroup} later.
+   *
+   * @param childKey A child key.
+   * @return The full key associated with the reserved slot.
+   */
+  public FULL_KEY reserve(CHILD_KEY childKey) {
+    while (true) {
+      FULL_KEY fullKey = groupManager.reserveNewSlot(childKey);
+      if (fullKey != null) {
+        return fullKey;
+      }
+      logger.debug(
+          "Failed to reserve a new value slot since the group was already closed. Retrying. key:{}",
+          childKey);
+    }
+  }
+
+  /**
+   * Marks the slot associated with the specified key READY and then, waits until the group which
+   * contains the slot is emitted.
+   *
+   * @param fullKey A full key associated with the slot already reserved with {@link
+   *     GroupCommitter#reserve}.
+   * @param value A value to be set to the slot.
+   * @throws GroupCommitException when group commit fails
+   */
+  public void ready(FULL_KEY fullKey, V value) throws GroupCommitException {
+    Keys<PARENT_KEY, CHILD_KEY, FULL_KEY> keys = keyManipulator.keysFromFullKey(fullKey);
+    boolean failed = false;
+    while (true) {
+      Group<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> group = groupManager.getGroup(keys);
+      if (group.putValueToSlotAndWait(keys.childKey, value)) {
+        return;
+      }
+      // Failing to put a value to the slot can happen only when the slot is moved from the original
+      // NormalGroup to a new DelayedGroup. So, only a single retry must be enough.
+      if (failed) {
+        throw new GroupCommitException(
+            String.format(
+                "Failed to put a value to the slot unexpectedly. Group:%s, FullKey:%s, Value:%s",
+                group, fullKey, value));
+      }
+      failed = true;
+      logger.debug(
+          "The state of the group has been changed. Retrying. Group:{}, Keys:{}", group, keys);
+    }
+  }
+
+  /**
+   * Removes the slot from the group.
+   *
+   * @param fullKey A full key to specify the slot.
+   */
+  public void remove(FULL_KEY fullKey) {
```

Comment:
When is this `remove` method intended to be called?

Reply:
Good question.

Let's say a transaction executes some CRUD operations, starts PREPARE, and fails before the group commit for coordinator state. In this case, a group commit slot is allocated for the transaction and needed to be removed by calling this method not to leave garbage slots.

Reply:
Updated the comment in https://github.com/scalar-labs/scalardb/pull/1607/commits/4ff928a8cf45c2e137c75fac93608e73f5b60040